akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala

```diff
@@ -4,46 +4,151 @@
+
+  override val shape = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[IOResult]) = {
+    val ioResultPromise = Promise[IOResult]()
+
+    val logic = new TimerGraphStageLogic(shape) with OutHandler {
+      val maxReadAhead = inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+
+      val bufferSize = if (maxReadAhead > 2) (chunkSize * maxReadAhead) / 2 else chunkSize
+      val buffer = ByteBuffer.allocate(bufferSize)
+
+      var channel: AsynchronousFileChannel = _
+      var position = 0L
+      var chunkCallback: Try[Int] ⇒ Unit = _
+      var readBytes = -1
+      var waitingForData = false
+
+      private def dataIsReady = readBytes != -1
+      private def setDataReady(bytes: Int) = readBytes = bytes
+
+      override def preStart(): Unit = {
+        try {
+          // this is a bit weird but required to keep existing semantics
+          require(Files.exists(path), s"Path '$path' does not exist")
+          require(Files.isRegularFile(path), s"Path '$path' is not a regular file")
+          require(Files.isReadable(path), s"Missing read permission for '$path'")
+        } catch {
+          case ex: IllegalArgumentException ⇒
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
+            throw ex
+        }
+        channel = AsynchronousFileChannel.open(path, StandardOpenOption.READ)
+
+        chunkCallback = getAsyncCallback[Try[Int]] {
+          case Success(readBytes) ⇒
+            waitingForData = false
+            setDataReady(readBytes)
+            if (isAvailable(out)) pushBuffer()
+
+          case Failure(ex) ⇒
+            failStage(ex)
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
+
+        }.invoke
+      }
+
+      private def pushBuffer(): Unit = {
+        if (readBytes > 0) {
+          buffer.flip()
+          val up = math.ceil(readBytes.toDouble / chunkSize).toInt
+          val list = for (_ ← 1 to up) yield {
+            val arr = new Array[Byte](if (buffer.remaining > chunkSize) chunkSize else buffer.remaining)
+            buffer.get(arr)
+            ByteString(arr)
+          }
+
+          if (readBytes == bufferSize)
+            emitMultiple(out, list)
+          else
+            emitMultiple(out, list.iterator, () ⇒ complete())
+
+          position += readBytes
+          buffer.clear()
+        } else {
+          // hit end, our work here is done
+          complete()
+        }
+        if (readBytes == bufferSize) {
+          setDataReady(-1)
+          read() //pre fetch
+        }
+      }
+
+      private def complete(): Unit = {
+        completeStage()
+        ioResultPromise.trySuccess(IOResult(position, Success(Done)))
+      }
+
+      private def read(): Unit = if (!waitingForData) {
+        channel.read(buffer, position, chunkCallback, FileSource.completionHandler)
```

Comment:
what happens if this throws? shouldn't we always make sure that ioResultPromise is completed?